### PR TITLE
Custom actions

### LINKED ISSUE #1206: Add option to generate custom actions in Admin UI
It would be handy to add some custom actions to Admin UI. Such actions that do not require any input and they just return a message and information whether the operation was succesful or not.

For example, for posts there could be a button Archive. This would be in the Admin UI in the view that shows the selected post.

A draft:

```javascript
var Post = new keystone.List('Post', {
    autokey: { path: 'slug', from: 'title', unique: true },
    map: { name: 'title' },
    defaultSort: '-createdAt',
    actions: {
       'Archive': archive
    }
});

var archive = function(next) {
    this.state = 'archived';
    this.save(function(err) {
        if (!err) {
            next(err, 'Post is archived');
        } else {
            next(err, 'Archiving failed');
        }
    });
}
```
Of course this example is not very useful, but I think that useful use cases exist.

### PR DESCRIPTION
**Pull request by adamscybot**
This pull request is to resolve #1206. Essentially it adds initial support for custom actions on item detail pages. I plan on extending this further to also support list pages. This lets you add extra buttons like this:

screen shot 2016-01-15 at 14 57 22

Usage:

``` javascript
Post.customAction('Archive', function(req, res, next) {
    var item = req.items[0];
    item.status = 'archived';
    item.save(function(err) {
        if (err) {
            throw new Error("Couldn't archive this post!");
        }

        next('Post was archived successfully!');
    })
})
```

You can optionally pass an options object that currently accepts `mobileText` for the button text on mobile. You can also override the unique slug assigned to the endpoint for the action with the `slug` option.

The item that the action was called on can be accessed via `req.items[0]`. I intend to go on to add custom actions to Lists, hence the array format. For lists, `req.items` will be an array of items that were "ticked" on the list page. This is going to require some UI work and probably support for things like batch delete to go with it for it to make sense. You will be able to limit it if you only want it on list view or you only want it on item view.

Errors (via throwing) and success messages (via next()) are pulled through to the UI as flash messages. Instead of utilising `Keystone.messages` bootstrapped into the template, I parse the asynchronous call response and add them dynamically with react. Please see #2090 for my reasoning behind this. This  work has also unlocked proper handling of messages of other asynchronous functions like delete which were marked as TODO. I will do a separate pull request for those. I hope this approach is suitable. It is my view that we should move to all error handling being parsed from responses to the calls and not from `Keystone.messages`. This will help down the line with true SPA support.

I'm happy with how it works on item views now, so if you're ok with it, it could be pulled in as WIP. I will then do the documentation.

One question I have is over positioning and styling. Is it ok as it is now? Or should we consolidate custom actions into a small dropdown button next to the save button. Just thinking of scenarios where someone has loads of custom actions.

### CODE REVIEW

**Comment by webteckie** [comment]
@adamscybot this looks very good!  This will help my app as well!  I vote for them being regular buttons as you currently have it.  And I think they should inherit the same style as the Save button.  To minimize these taking extra real estate in the screen, perhaps the `reset changes` should just become a `Reset` button and `delete <list>` should become a `Delete` button both with a title tooltip with the extra info.

cc: @jossmac

**Comment by JedWatson** [comment]
@adamscybot I **really** love what you've done here. I've been planning to add something like this in a slightly different way for 0.4 (specifically, exposing methods on a model but configuring the access to them client-side, not server-side through an Admin UI plugin API)

I've also been holding off on adding something like this while we explore all the possible ways of implementing custom actions for list items, so that we don't go down one path then have to back-track and break existing customisations.

_however_ I like your approach so much I think it might be worth going with regardless.

Positioning and styling are great for this variant; You might notice we've left room in the sidebar (could be a callout menu on mobile) which could also be used for custom components / actions. We may also want to make actions available at the top of the edit form.

One thing I've been thinking about that hasn't been taken into account here is the availability of custom actions; e.g. if a post is archived, it doesn't make sense to still have an "archive" button.

I'm going to think some more on this before merging, ping @creynders and @ericelliott as well for thoughts.

@adamscybot if you haven't seen it, you might also be interested in watching my nodevember talk from last year in which I outline some of the ways we're planning to open up Keystone for extensions, it's pretty relevant to this: https://www.youtube.com/watch?v=of2VM3Wwkxk

**Comment by webteckie** [comment]
@adamscybot @JedWatson what about adding a o dependsOn config here?  With all that power @snowkeeper   added to dependsOn it shouldn't be a problem?

**Comment by creynders** [comment]
Looks cool. Definitely need to be able to disable specific action buttons though. I'd go for a 'guarded' approach here though, i.e. call a handler per action/button type which returns a boolean to determine the `disabled` state (and not e.g. an object with booleans for each action type).

Then, I'm not fond of `callback` as the property name, since it's a bit of a misnomer IMO, wouldn't `action` be more appropriate? .

One more thing, I'm not entirely sure about the URL. I think I'd prefer `/api/<list>/<id>/actions/<action>`
Oh, and then I'd change the `delete` action URL too to `/api/<list>/<id>/actions/delete`

**Comment by creynders** [comment]
@adamscybot wow, that's a truly great PR. From start to finish. Totally :+1: as far as I'm concerned.
Just one minor thing we need to consider: whether `save`'s a good property name. Since, if I'm not mistaken, the item is _not_ saved after the custom action is triggered (only before), right? The name might be confusing then. How 'bout `preSave`? Or it could be made more versatile, something like: 
```
{
  save: 'pre'
}
```
with possible values `'pre'`, `'post'` or `['pre', 'post']`?
Or even:
```js
{
  save: { pre: true, post: false}
}
```
What do you think?

**Comment by webteckie** [comment]
@adamscybot looking good!  Can't wait for this :-)  As I was reading your comment the same impression as @creynders  about save crossed my mind.  Thus, I agree with his suggestions and prefer the very last one.  Thanks!

File: admin/client/components/EditForm.js [hunk 1]
```diff
@@ -65,12 +68,35 @@ var EditForm = React.createClass({
 		list.deleteItem(data.id, err => {
 			if (err) {
 				console.error(`Problem deleting ${list.singular}: ${data.name}`);
-				// TODO: slow a flash message on form
+				// TODO: show a flash message on form
```
Comment (ericelliott) [inline admin/client/components/EditForm.js]: Is there an issue tracking this TODO, or is it just hiding here in the code? =)

**Comment by ericelliott** [comment]
Is there some corresponding documentation for the new functionality and API endpoints?

**Comment by ttsirkia** [comment]
Very nice job @adamscybot ! I had almost forgotten that I had created that issue. This implements exactly the feature that I was looking for. Now I have a few use cases for this: For example, I have a model for email validations. If there is a need to resend the validation message, this feature allows to implement that feature into the Admin UI.

**Comment by morenoh149** [comment]
@adamscybot https://github.com/keystonejs/keystone/issues/2151

**Comment by webteckie** [comment]
@JedWatson can we please take some action on this one?  I think this is a pretty useful feature!

**Comment by VinayaSathyanarayana** [comment]
When is this getting merged?

**Comment by creynders** [comment]
:100: times :+1: Let's get this merged!

**Comment by webteckie** [comment]
:+1: :+1:

**Comment by creynders** [comment]
@JedWatson could we get your blessing on this one? If so, @adamscybot could you fix the merge conflicts? Sorry for the long wait!

**Comment by webteckie** [comment]
@creynders @adamscybot I discussed this one with @JedWatson and he thinks it is best to address it after 0.4 since otherwise it will make 0.4 harder to get out, which is his first keystonejs priority. Trust me I want this as bad as you all do:-)  I sympathize with Jed and to help get to this faster our efforts at this moment would be better spent helping with getting 0.4 out ASAP.  I'm pretty sure this PR will be amongst the first to be considered once 0.4 is published.  Now if Jed can change his mind about including this in now then that's something else :-)

**Comment by mxstbr** [comment]
I'm not entirely up to speed here, why would we wait to get this in until the PR is out of date and hard to rebase instead of now? Is there any more work that'd need to be done for 0.4 to be released, how would this block that?

**Comment by webteckie** [comment]
@mxstbr there's a few of us that really want this.  If you can convince @JedWatson to merge it that would be great.  I had a talk about it with him and there were concerns of merging in and how it would affect the remaining of the 0.4 work.  If that's no longer a concern then it should definitely be merged!

**Comment by SharadKumar** [comment]
Hi @adamscybot - @JedWatson and team are preparing v0.4 beta. Are you plan to rebase since there has been quite a lot of changes since then and bring this up to date?

Appreciate all the effort mate.

**Comment by VinayaSathyanarayana** [comment]
Any updates on this?

**Comment by michaek** [comment]
This is a great feature addition. If there's any way I can help prep this to go into KeystoneJS, I would be happy to. It seems like @adamscybot is waiting on @JedWatson's answer on whether the internal API is stable enough to re-roll this. As 4.0 is now in beta, does that make it more likely to be stable?

**Comment by VinayaSathyanarayana** [comment]
Any updates on this PR getting merged?

**Comment by VinayaSathyanarayana** [comment]
Any updates on this PR getting merged?

**Comment by VinayaSathyanarayana** [comment]
Any updates on this PR getting merged?

**Comment by VinayaSathyanarayana** [comment]
Any updates on this PR getting merged?

**Comment by ichiriac** [comment]
any news ? I also need this PR, exactly the same need

**Comment by Twansparant** [comment]
This functionality would be great yes! +1

**Comment by JedWatson** [comment]
This isn't going to land in v4, but is in scope for v5 (see keystonejs/keystone-5#1002)

**Pull request closed without merging**